pages/glossary/label-in-name-algorithm.md

```diff
@@ -0,0 +1,39 @@
+---
+title: Label in Name Algorithm
+key: label-in-name-algorithm
+unambiguous: true
+objective: false
+input_aspects:
+  - CSS styling
+  - DOM tree
+---
+
+To check whether an [element][] has <dfn>its label contained in its name</dfn>, follow this algorithm:
+
+Let 'label' be the [visible inner text][] of the target element.  Let 'name' be the [accessible name][] of the target element.  Both 'label' and 'name' are strings.
+
+Sub-algorithm to tokenize a string:
+
+- Convert the string to lower case.
+- For each character that either a) represents non-text content, or b) isn't a letter or a digit: replace that character with a space character.
+    - For a) Judgment of "non-text" probably can't be fully automated.  For example: "X" for "close" probably can be automated, but presumably there are more cases than this.
+    - For b) Use the Unicode classes Letter, Mark, and "Number, Decimal Digit [Nd]". (This will exclude hyphens, punctuation, emoji, and more.)
+- Remove all characters that are within parentheses (AKA round brackets).
+    - Ignore square brackets and braces.
+- Split the string into a list of strings, using a greedy [whitespace][] regular expression as the separator.
```

Comment:
Good call out.  I'm going to keep defending regular expressions for the time being, because I can't see how to handle this case otherwise: `<a href="#" aria-label="Discover Italy">Discover It</a>`. 

Obviously we want to avoid a situation like "[over sixty seconds to match a 29-character string](https://swtch.com/~rsc/regexp/regexp1.html)".  It seems that the article's concern only applies to regular expressions which involve backtracking, and my algorithm's regex ("a greedy whitespace regular expression") isn't one of those.  In Javascript, my algorithm's regex would be this: `const regex = /\p{White_Space}+/gu;`, which (I think) can't result in any backtracking, because there is nothing before the "atom" `\p{White_Space}` to backtrack to.  

As for the concern about O(m * n^2) complexity, I'm unclear on what exactly m and n are, but I'm open to doing some tests.   As always with Big O notation, even if the scaling (for large m and n) is bad, the unspoken constant "C" might be small enough that it's not a real problem for our uses cases.  

Regarding Unicode normalization: no argument there.  I'll try to add something to that end.  Not sure when the best time to do that would be.

Reply:
Thinking about this a bit more, a more serious issue is RegEx and my suggested algorithm only work for a restricted set of languages. Thai, Chinese and most Japanese text don't use whitespace to separate words.

MDN also recommends against using Regex for word breaking:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Word_boundary_assertion#examples

One way of resolving this - instead of: 

> Split the string into a list of strings, using a greedy [whitespace][] regular expression as the separator.

say 

> Split the string into a list of strings, using the word breaking rules for the inherited programmatic language.

This wouldn't force a change on implementations already using regexes, but means the rule is less likely to need changed to support Thai, Chinese etc. 

Of note: CSS doesn't specify exactly where word boundaries occur in different languages but does link to additional specs that do: https://drafts.csswg.org/css-text/#soft-wrap-opportunity

Reply:
Sorry I haven't had time to work on this.   I think I will later this month.